### ISSUE TITLE
Relax Pixel unit constraint

### ISSUE THREAD

**Issue by MartinSStewart** [ianmackenzie/elm-geometry-svg#14]
The SVG functions all require that the `elm-geometry` type they are being constructed from have the Pixel unit. While upgrading a project to v2 of this package, I found this to be a problem.

I'm drawing a map. The objects on the map can be thought of as being in world coordinates with Length units. I'd like to convert those objects to SVG and then use `Svg.scaleBy` and `Svg.translateBy` to convert to screen coordinates. Since `elm-geometry-svg` requires Pixels though, it seems to be requiring that I convert from world coordinates to screen coordinates first and then draw my SVG. The problem is that transforming lots of elm-geometry values is a lot less efficient than just updating an `Svg.scaleBy` and `Svg.translateBy`.

I think it would be best if `elm-geometry-svg` just allowed for any unit to be used when drawing SVG.

**Comment by ianmackenzie** [comment]
Yeah, I think you're probably right. I'm actually working on an `elm-2d-drawing` package (roughly speaking a high level interface to SVG) that handles this much more explicitly - every drawing element has its units/coordinate system tracked at compile time (just like `elm-geometry` values themselves), which gives you the best of both worlds. You can create a drawing element in meters, and then do something like

```elm
worldDrawingElement |> Drawing2d.at screenResolution
```

to convert to a drawing element in pixels all in a nicely type-checked way. Under the hood this will correspond to just applying an SVG scale, which as you point out is much more efficient than transforming the actual geometry.

In raw SVG there's no nice way to track this sort of thing - for example, having `elm-geometry-svg` change to accept any kind of units would mean that this would be legal:

```elm
Svg.lineSegment2d [] (LineSegment.from (Point2d.meters 10 10) (Point2d.meters 20 20))
    |> Svg.translateBy (Vector2d.pixels 300 400)
```

I think maybe this makes sense though - `elm-geometry-svg` is then a pretty low-level, slightly 'unsafe' package and hopefully `elm-2d-drawing` will be the safer, more expressive, higher-level package that's more commonly used.

I'll think about this a bit more today but I'm leaning towards relaxing the units restriction when I publish an updated version of `elm-geometry-svg` that uses `elm-geometry` 3.x.

Thanks for being patient and helping me work through these design issues!

**Comment by MartinSStewart** [comment]
Yeah, the `elm-2d-drawing` approach seems like the best way to go about this.

> Thanks for being patient and helping me work through these design issues!

Thanks for making these packages in the first place!

**Comment by ianmackenzie** [comment]
OK 3.0 is now published with the `Pixels` constraint removed - I also added `Svg.at` and `Svg.at_` functions for unit scaling (which work like the corresponding functions in `elm-units` and `elm-geometry`). And I've added [some notes about unit conversions](https://package.elm-lang.org/packages/ianmackenzie/elm-geometry-svg/3.0.0/Geometry-Svg#a-note-on-unit-conversions) to the documentation - if you could take a look at some point and see if they make sense, that would be great.

**Comment by MartinSStewart** [comment]
I've read through it and it makes sense to me. I'm glad you mentioned that the underlying units matter some cases such as in your example here.
```
Svg.lineSegment2d [] <|
    LineSegment2d.from
        (Point2d.inches 1 2)
        (Point2d.inches 3 4)
```
That's tripped me up before when dealing with `Quantity.round` so I think it's good to make the user mindful of those situations.

**Comment by ianmackenzie** [comment]
Yeah, I just realized the underlying units issue as I was writing some of the examples for the documentation! I think the only alternative would be to require every function to take a unit conversion function as the first argument, e.g.

```elm
lineSegment2d : 
    (Quantity Float units -> Float) 
    -> List (Attribute msg)
    -> LineSegment2d units coordinates
    -> Svg msg
```

which you could then call as

```elm
Svg.lineSegment2d Length.inInches [] <|
    LineSegment2d.from
        (Point2d.inches 1 2)
        (Point2d.inches 3 4)
```

if you really wanted the resulting SVG to be in inch units - but that seemed annoying, and encouraging an explicit unit conversion step in these cases (by applying a scale factor to the SVG) seemed like a better, more lightweight approach.

What sort of situations tripped you up with `Quantity.round`? I did try to add [some warnings](https://package.elm-lang.org/packages/ianmackenzie/elm-units/latest/Quantity#-int-float-conversion) to the docs about that issue - do you think it would be useful for the individual functions like `round` and `floor` to link back to those warnings with something like a "See caveats" note? Or are there other issues with using `round` that aren't covered by the warnings?

**Comment by MartinSStewart** [comment]
Yeah, I think it would be good to provide a link in those functions to that warning. I missed it because I usually briefly look at the readme for a quick overview. Then just look at the function specific documentation as I'm writing code (the intellij Elm plugin makes it easy to view the documentation for individual functions).

In my case what happened is that I was upgrading `elm-geometry`. I was going through code for a game and replacing things like `Point2d.fromCoordinates (5, 5)` with `Point2d.xy (Length.centimeters 5) (Length.centimeters 5)`. The game uses a grid and I have a lot of code that snaps various entities to the grid with `Basics.round` which I upgraded to `Quantity.round`.

**Comment by ianmackenzie** [comment]
OK, just added in some explicit links and some extra clarification about how rounding works on `Quantity` values, so that will be in the next release of `elm-units`. Do you think it would be useful to have something like

```elm
Quantity.roundToMultipleOf : 
    Quantity Float units
    -> Quantity Float units
    -> Quantity Float units
```

? You'd use it as

```elm
Quantity.roundToMultipleOf (Length.centimeters 1) (Length.millimeters 123)
--> Length.centimeters 12
```

(Hopefully we could come up a with a shorter name that was still reasonably clear...maybe `snapTo`?)
```

**Comment by ianmackenzie** [comment]
Thanks for the experience report on the upgrade attempt, and I'm sorry it didn't work out in that case - I'm not _entirely_ surprised, since 1.x -> 3.x is definitely a big shift, but if there's anything you think I could do to `elm-geometry` (or `elm-units`) to make that process easier for others then I'd love to hear it.

**Comment by MartinSStewart** [comment]
```
Quantity.roundToMultipleOf : 
    Quantity Float units
    -> Quantity Float units
    -> Quantity Float units
```
In my case this wouldn't work. I need to get `Quantity Int units` back because I'm working with data that can only be on integer grid points. My work-around was to create a `LengthUnit` type with two functions `length: number -> Quantity number LengthUnit` and `inLength: Quantity number LengthUnit -> number` that would internally represent the values exactly as I passed them in.

> if there's anything you think I could do to elm-geometry (or elm-units) to make that process easier for others then I'd love to hear it.

While upgrading, I think a lot of my time was spent double checking that I hadn't gotten the argument order wrong when converting `a - b` into `a |> Quantity.minus b` or `a > b` into `a |> Quantity.greaterThan b`. Sometimes I needed to introduce let...in or additional functions because what could previously fit into one line became too long when using Quantity functions (though in hindsight, given how frequently Quantity is used, I should have aliased it to Q to make it less verbose).

These are minor things but if I had perhaps a 1% chance of introducing a logic bug every time I did one of these, doing this hundreds of times over a 20k project meant I was probably going to end up with a few of logic bugs which the compiler can't catch. I guess this could be improved with some kind of upgrade tool that automates stuff like `a - b` ->  `a |> Quantity.minus b` ? I'm not sure how easy it is to do that but my gut feeling is that it wouldn't be worth the effort.

Additionally, I also was spending a lot of time upgrading my adhoc geometry library for Point2i, Vector2i, and BoundingBox2i. I'm not sure how many other people have projects that use `elm-geometry` and also have a lot of grid based interactions but I guess the solution here would be to introduce those types into elm-geometry. I don't know how to make a nice API around this though given the aforementioned Quantity.round issue.

Anyway, I don't want to make it should like these problems were insurmountable, but for a project that's nearing completion it did not feel worthwhile to continue the upgrade. The work you've done to build up `elm-geometry` and all the supporting packages is very much appreciated!
